dll/win32/syssetup/install.c

```diff
@@ -524,24 +526,177 @@ EnableUserModePnpManager(VOID)
+                                                DEVICE_NOTIFY_WINDOW_HANDLE | DEVICE_NOTIFY_ALL_INTERFACE_CLASSES);
+
+    return *hDeviceNotify != NULL;
+}
+
+static BOOLEAN
+GetFriendlyDeviceName(
+    IN PDEV_BROADCAST_DEVICEINTERFACE pDevice,
+    OUT LPCWSTR* pszName)
+{
+    BOOLEAN result = FALSE, DeviceInterfaceDataInitialized = FALSE;
+    HDEVINFO hList = NULL;
+    SP_DEVICE_INTERFACE_DATA DeviceInterfaceData = { 0 };
+
+    if (!pDevice
+        || ~pDevice->dbcc_devicetype & DBT_DEVTYP_DEVICEINTERFACE
+        || !pDevice->dbcc_name)
+    {
+        DPRINT1("pDevice is not initialized properly\n");
+        goto cleanup;
+    }
+
+    /* Create an empty device info set */
+    hList = SetupDiCreateDeviceInfoList(NULL, 0);
+    if (!hList)
+    {
+        DPRINT1("SetupDiCreateDeviceInfoList failed (error %d)\n", GetLastError());
+        goto cleanup;
+    }
+
+    /* Set up DeviceInterfaceData */
+    DeviceInterfaceData.cbSize = sizeof(SP_DEVICE_INTERFACE_DATA);
+
+    /* Open device interface */
+    DeviceInterfaceDataInitialized = SetupDiOpenDeviceInterface(hList,
+                                                                &pDevice->dbcc_name[0],
+                                                                0,
+                                                                &DeviceInterfaceData);
+    if (!DeviceInterfaceDataInitialized)
+    {
+        DPRINT1("SetupDiOpenDeviceInterface failed (error %d)\n", GetLastError());
+        goto cleanup;
+    }
+
+    for (INT index = 0; ; index++)
+    {
+        SP_DEVINFO_DATA devInfo = { 0 };
+
+        /* Variables for SetupDiGetDeviceRegistryProperty */
+        TCHAR szName[1024];
+        DWORD PropertyRegDataType = 0, RequiredSize = 0;
+
+        /* Init devInfo */
+        devInfo.cbSize = sizeof(SP_DEVINFO_DATA);
+
+        /* Enumerate through hList */
+        if (SetupDiEnumDeviceInfo(hList, index, &devInfo))
+        {
+            /* Get friendly name of devInfo */
+            result = SetupDiGetDeviceRegistryProperty(hList,
+                                                      &devInfo,
+                                                      SPDRP_FRIENDLYNAME,
+                                                      &PropertyRegDataType,
+                                                      (BYTE*)szName,
+                                                      sizeof(szName),
+                                                      &RequiredSize);
+
+            if (result)
+            {
+                /* We found a name - allocate buffer, break the loop */
+                *pszName = (LPWSTR)HeapAlloc(GetProcessHeap(), 0, sizeof(szName));
+                if (!*pszName)
+                {
+                    DPRINT1("HeapAlloc failed\n");
+                    result = FALSE;
+                    goto cleanup;
+                }
+
+                CopyMemory((LPWSTR)*pszName, szName, sizeof(TCHAR) + 1024);
```

Comment:
I think you meant `*` here rather than `+`; though `RequiredSize` would be even better.